File: Py/breakpoints.py
# Step 1: Install required packages (no need in Python)
!pip install rpy2

# 1.1 Load the required packages
import yfinance as yf
import pandas as pd
import matplotlib.pyplot as plt

# 1.2 Load rpy2 package and set up a connection to an R session:
import rpy2.robjects.packages as rpackages
import rpy2.robjects as robjects

# 1.3 Install strucchange in R
utils = rpackages.importr('utils')
utils.install_packages('strucchange')

# 1.4 Activate strucchange
strucchange = rpackages.importr('strucchange')


# Step 2: Retrieve the last 5 years of data for AMD
amd = yf.download("AMD", start=pd.Timestamp.today() - pd.DateOffset(years=5), end=pd.Timestamp.today())

# 2.1 Extract the adjusted close prices as a time series
amd_ts = amd['Adj Close']

# Step 3: Perform the breakpoint test
bp_test = breakpoints(amd_ts, h=1)

# 3.1 Extract the estimated breakpoints
bp = bp_test.breakpoints

# Step 4: Calculate means of each breakpoint segment
# 4.1 Calculate the first mean
means = [amd_ts[0:bp[0]].mean()]

# 4.2 Calculate the rest of the means between breakpoints
for i in range(len(bp)):
    if i == len(bp)-1:
        means.append(amd_ts[(bp[i]+1):].mean())
    else:
        means.append(amd_ts[(bp[i]+1):bp[i+1]].mean())

# Step 5: Plot time series with vertical lines at breakpoints and horizontal lines at mean values
# 5.1 Plot the original time series
plt.plot(amd_ts.index, amd_ts.values)
plt.title("AMD Adjusted Close Prices")

# 5.2 Plot the breakpoints
for b in bp:
    plt.axvline(x=amd_ts.index[b], color='blue')

# Plot means between breakpoints
for i in range(len(means)):
    if i == 0:
        plt.axhline(y=means[i], xmin=0, xmax=bp[i]/len(amd_ts), color='red', linestyle='--')
    elif i == len(means)-1:
        plt.axhline(y=means[i], xmin=bp[i-1]/len(amd_ts), xmax=1, color='red', linestyle='--')
    else:
        plt.axhline(y=means[i], xmin=bp[i-1]/len(amd_ts), xmax=bp[i]/len(amd_ts), color='red', linestyle='--')

# Show the plot
plt.show()

File: R/breakpoints.R
# Step 1: Install required packages
install.packages("strucchange")

# 1.1 Load the required packages
library(tidyquant)
library(strucchange)

# Step 2: Retrieve the last 5 years of data for AMD
amd <- tq_get("AMD", from = Sys.Date() - years(5), to = Sys.Date())

# 2.1 Extract the adjusted close prices as a time series
amd_ts <- ts(amd$adjusted, frequency = 1)

# Step 3: Perform the breakpoitns test test
bp_test <- breakpoints(amd_ts ~ 1)

# 3.1 Extract the estimated breakpoints
bp <- bp_test$breakpoints


# Step 4:  Calculate means of each breakpoint segment
# 4.1 Calculate the first mean
means <- c(mean(amd_ts[1:bp[1]]))

# 4.2 Calculate the rest of the means between breakpoints
for (i in 1:length(bp)) {
        if (i == length(bp)) {
                means <- c(means, mean(amd_ts[(bp[i] + 1):length(amd_ts)]))
        } else {
                means <- c(means, mean(amd_ts[(bp[i] + 1):bp[i + 1]]))
        }
}



# Step 5: Plot time series with vertical lines at breakpoints and horizontal lines at mean values
# 5.1 Plot the original time series
plot(amd_ts, main = "AMD Adjusted Close Prices", ylab = "AMD Price", yaxt = "n")

# 5.2 Plot the breakpoints
abline(v = bp, col = "blue")

# Plot means between breakpoints
for (i in 1:length(means)) {
        if (i == 1) {
                lines(c(1, bp[i]), c(means[i], means[i]), col = "red", lty = 2)
        } else if (i == length(means)) {
                lines(c(bp[i - 1] + 1, length(amd_ts)), c(means[i], means[i]), col = "red", lty = 2)
        } else {
                lines(c(bp[i - 1] + 1, bp[i]), c(means[i], means[i]), col = "red", lty = 2)
        }
}
# Add gridlines
grid()

## END


##BONUS: to prettify the standard R plot, not ussing ggplot2
#add this code between Step 5.1 + 5.2

legend("topleft", legend = "Breakpoints", col = "blue", lty = 1, bty = "n")
legend("topright", legend = "Segment Means", col = "red", lty = 2, bty = "n")

# Set y-axis intervals in steps of 20
yintervals <- pretty(amd_ts, n = ceiling((max(amd_ts) - min(amd_ts))/20))
axis(2, at = yintervals)
